azure-iot-device/tests/iothub/pipeline/test_iothub_pipeline.py

```diff
@@ -7,505 +7,1407 @@
 import pytest
 import logging
 import six.moves.urllib as urllib
+from azure.iot.device.common.pipeline import (
+    pipeline_stages_base,
+    pipeline_stages_mqtt,
+    pipeline_ops_base,
+)
+from azure.iot.device.iothub.pipeline import (
+    pipeline_stages_iothub,
+    pipeline_stages_iothub_mqtt,
+    pipeline_ops_iothub,
+    pipeline_events_iothub,
+)
 from azure.iot.device.iothub import Message
 from azure.iot.device.iothub.pipeline import IoTHubPipeline, constant
-from azure.iot.device.iothub.auth import SymmetricKeyAuthenticationProvider
-from mock import MagicMock, patch, ANY
-from datetime import date
-
-connection_string_format = "HostName={};DeviceId={};SharedAccessKey={}"
-connection_string_format_mod = "HostName={};DeviceId={};ModuleId={};SharedAccessKey={}"
-fake_shared_access_key = "Zm9vYmFy"
-fake_hostname = "beauxbatons.academy-net"
-fake_device_id = "MyPensieve"
-fake_module_id = "MemoryCharms"
-fake_event = "Wingardian Leviosa"
-fake_event_2 = fake_event + " again!"
-
-before_sys_key = "%24."
-after_sys_key = "="
-topic_separator = "&"
-message_id_key = "mid"
-fake_message_id = "spell-1234"
-custom_property_value = "yes"
-custom_property_name = "dementor_alert"
-fake_topic = "devices/" + fake_device_id + "/messages/events/"
-encoded_fake_topic = (
-    fake_topic
-    + before_sys_key
-    + message_id_key
-    + after_sys_key
-    + fake_message_id
-    + topic_separator
-    + custom_property_name
-    + after_sys_key
-    + custom_property_value
-)
-subscribe_input_message_topic = (
-    "devices/" + fake_device_id + "/modules/" + fake_module_id + "/inputs/#"
+from azure.iot.device.iothub.auth import (
+    SymmetricKeyAuthenticationProvider,
+    X509AuthenticationProvider,
 )
 
-subscribe_c2d_topic = "devices/" + fake_device_id + "/messages/devicebound/#"
-subscribe_methods_topic = "$iothub/methods/POST/#"
 
-send_msg_qos = 1
+# Update this list with features as they are added to the SDK
+all_features = [
+    constant.C2D_MSG,
+    constant.INPUT_MSG,
+    constant.METHODS,
+    constant.TWIN,
+    constant.TWIN_PATCHES,
+]
+
+
+@pytest.fixture
+def auth_provider(mocker):
```

Comment:
>auth_provider [](start = 4, length = 13)

this fixture is specific to this class ?

Reply:
Yes. Depending on the level of abstraction, auth_provider is handled differently in testing. In this test module, which was entirely agnostic re: auth providers (up until your latest PR of course), I just used a mock because it didn't matter what it was. In other test modules, it needs to be more concrete.

---
In reply to: [299253865](https://github.com/Azure/azure-iot-sdk-python-preview/pull/120#discussion_r299253865) [](ancestors = 299253865)